### DISCUSSION TITLE
Grad All Zero

### DISCUSSION THREAD

**Discussion by zichenwang01** [mitsuba-renderer/mitsuba3#811] [Q&A] [answered]
Hi,

I coded a custom integrator, defined a loss function between the reference image and the custom rendered image, and want to backpropagate the loss to the custom rendered image. However, I am having the grads all zero right now.

    image1 = mi.render(scene, spp=64)

    sdf_integrator = mi.load_dict({
       'type': 'sdf-direct-ps'
    })
    image2 = sdf_integrator.render(scene=scene, sdf=sdf, sensor=sensor, spp=64)

    dr.enable_grad(image1)
    dr.enable_grad(image2)
    loss = l1_loss(image1, image2)
    # dr.backward(loss)
    dr.backward_to(image2)
    grad = dr.grad(image2)

The weird place is that if I backpropagate to image1, then everything is fine, but things do not work when I backpropagate to image2.

Thanks for any help in advance!

**Comment by zichenwang01** [comment, marked as answer]
Nvm, I think I figured out the problem: I need to `dr.set_grad(loss, 1)` before backpropagating!
